hotspot/main.go

```diff
@@ -0,0 +1,259 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+
+// The hot spot example program is a read/write workload intended to always hit
+// the exact same row. It performs reads and writes to simulate a super
+// contentious load.
+
+package main
+
+import (
+	"database/sql"
+	"flag"
+	"fmt"
+	"log"
+	"math/rand"
+	"net/url"
+	"os"
+	"os/signal"
+	"runtime"
+	"sync"
+	"sync/atomic"
+	"syscall"
+	"time"
+
+	// Import postgres driver.
+	_ "github.com/lib/pq"
+)
+
+const (
+	createDatabaseStatement = "CREATE DATABASE IF NOT EXISTS hot"
+	createTableStatement    = `
+	CREATE TABLE IF NOT EXISTS hot.spot (
+	  id BIGINT NOT NULL,
+	  value BIGINT,
+	  PRIMARY KEY (id)
+	)`
+)
+
+var concurrency = flag.Int("concurrency", 2*runtime.NumCPU(), "Number of concurrent reading/writing processes")
+var writePercent = flag.Int("writePercent", 50, "Percentage, from 0 to 100 of the operations that will perform writes instead of reads")
+var tolerateErrors = flag.Bool("tolerate-errors", false, "Keep running on error")
+var outputInterval = flag.Duration("output-interval", 1*time.Second, "Interval of output")
+var duration = flag.Duration("duration", 0, "The duration to run. If 0, run forever.")
+var benchmarkName = flag.String("benchmark-name", "BenchmarkHotSpot", "Test name to report for Go benchmark results.")
+
+var readCount, writeCount uint64
+
+// A hotSpotWriter writes and reads values from one row in an infinite loop.
+type hotSpotWriter struct {
+	db   *sql.DB
+	rand *rand.Rand
+}
+
+func newHotSpotWriter(db *sql.DB) hotSpotWriter {
+	source := rand.NewSource(int64(time.Now().UnixNano()))
+	return hotSpotWriter{
+		db:   db,
+		rand: rand.New(source),
+	}
+}
+
+// run is an infinite loop in which the hotSpotWriter continuously attempts to
+// read and write values from a single row.
+func (w hotSpotWriter) run(errCh chan<- error, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	wPercent := *writePercent
+	for {
+		if w.rand.Intn(100) < wPercent {
+			if _, err := w.db.Exec("UPSERT INTO hot.spot(id, value) VALUES (1, $1)", rand.Int63()); err != nil {
+				errCh <- err
+			} else {
+				atomic.AddUint64(&writeCount, 1)
+			}
+		} else {
+			var value int
+			if err := w.db.QueryRow("SELECT value FROM hot.spot WHERE id = 1").Scan(&value); err != nil {
```

Comment:
What happens here if the `SELECT` happens on an empty table? I think some sort of error is returned, but it is expected in that case, right?